File: Venkatesh/Topological-Sorting/Dependencies.py
def sequence(dependencies, n):
    graph = [[] for _ in range(n)]
    indegree = [0] * n
    
    for u, v in dependencies:
        graph[u].append(v)
        indegree[v] += 1

    queue = []
    for i in range(n):
        if indegree[i] == 0:
            queue.append(i)
    result = []

    while queue:
        u = queue.pop(0)
        result.append(u)
        for v in graph[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                queue.append(v)

    return result

n = 6
dependencies = [(5, 2), (5, 0), (4, 0), (4, 1), (2, 3), (3, 1)]

print(sequence(dependencies, n))

File: Venkatesh/Topological-Sorting/Dependencies.java
import java.util.*;
import java.util.LinkedList;

public class Dependencies {
    public static List<Integer> sequence(int[][] dependencies, int n) {
        List<Integer> result = new ArrayList<>();
        if (n < 0) return result;
        
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[n];

        for (int i = 0; i < n; i++) {
            graph.put(i, new ArrayList<>());
        }
        for (int[] dependency : dependencies) {
            graph.get(dependency[0]).add(dependency[1]);
            inDegree[dependency[1]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            int node = queue.poll();
            result.add(node);
            for (int neighbor : graph.get(node)) {
                inDegree[neighbor]--;
                if (inDegree[neighbor] == 0) {
                    queue.offer(neighbor);
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int n = 6;
        int[][] dependencies = {{5, 2}, {5, 0}, {4, 0}, {4, 1}, {2, 3}, {3, 1}};

        System.out.println(sequence(dependencies, n));
    }
}